README.rst
Twisted GitHub Service Hooks Server
===================================


A picture is worth a thousand words:

.. figure:: http://adi.roiban.ro/media/img/articles/2013/txghserf.jpg

This is just a boilerplate for implementing a GitHub Service Hooks server
based on Klein and Twisted.

Implement your on "run.py" and start it like::

    twistd -n web --class=run.resource

It will parse incoming hooks and return an 'Event' containing hook name,
event name and JSON payload.

Configure GitHub Hooks using::

    http://host:port/hook/HOOK-NAME

HOOK-NAME is just a random string to track the source of the hook or
implement multiple hooks on the same server.

It also include a simple web tool to register hooks or check hook status.
Login using GitHub credentials::

    http://host:port/

..  warning::
    The JS gui will work fine from localhost, since GitHub API accept all
    localhost origins.
    To allow the JS gui to connect to GitHub from a public server, you will
    need to register the server address as on GitHub OAuth application.


Development
-----------

Get virtual environment and install dependencies::

    make deps

Run the tests::

    make test

Run the server::

    make run


TODO
----

* Add ssl security check instead of IP filter.
* Add proper pagination for GitHub API... not it just request 1000 entries.
* Add configuration and checking of secret


Thanks
------

This project is standing on the shoulders of giants:

* Twisted Matrix
* Klein
* AngularJS
* AngujarJS-UI Bootstrap
* Brandon Sterne for cidr.py


Changelog
0.3.0 (2013-03-06)
==================

* Filter IP based on CIDR blocks.


0.2.3 (2013-03-04)
==================

* Add Credentials to XMLHttpRequest.


0.2.2 (2013-03-04)
==================

* Fix Python package definition to include partials HTML files.


0.2.1 (2013-03-04)
==================

* Fix Python package definition to include files from lib.


0.2.0 (2013-03-04)
==================

* Add GUI for managing GitHub hooks.


0.1.9 (2013-03-01)
==================

* Decode contend based on content-type header.


0.1.8 (2013-03-01)
==================

* Improve error reporting.


0.1.7 (2013-03-01)
==================

* Fix parsing 'push' events as 'application/x-www-form-urlencoded'.


0.1.6 (2013-03-01)
==================

* List received input in error log message.


0.1.5 (2013-03-01)
==================

* Update list of GitHub IPs.


0.1.4 (2013-02-28)
==================

* Initial release.

txghserf/tests/test_server.py
"""
Tests for requests dispatcher.

Functional test using curl:

 curl --data '{"param1": "test"}' \
    -H "X-Github-Event: issue_comment" \
    -H "Content-type: application/json"\
    http://localhost:8080/hook/test

curl --data 'payload=%7B%22key%22%3A%22value%22%7D' \
    -H "X-Github-Event: push" \
    -H "Content-type: application/x-www-form-urlencoded" \
     http://localhost:8080/hook/something

"""
from StringIO import StringIO

from twisted.internet.address import IPv4Address
from twisted.trial.unittest import TestCase
from twisted.web.test.test_web import DummyRequest

from txghserf.server import CONFIGURATION, expand_allowed_ips, hook


class TestServer(TestCase):
    """
    Tests for resource handlers.
    """

    def setUp(self):
        super(TestServer, self).setUp()
        self._allow_cidr = CONFIGURATION['allow_cidr'][:]
        self._callback = CONFIGURATION['callback']

        self.request = DummyRequest([])
        ip = '10.0.0.100'
        CONFIGURATION['allow_cidr'] = ['10.0.0.0/24']
        expand_allowed_ips()
        self.request.client = IPv4Address('TCP', ip, 2345)

    def tearDown(self):
        CONFIGURATION['allow_cidr'] = self._allow_cidr
        expand_allowed_ips()
        CONFIGURATION['callback'] = self._callback
        super(TestServer, self).tearDown()

    def test_allow_cidr_change_runtime(self):
        """
        When allow_cidr is changed at runtime, `expand_allowed_ips` must
        be called to update the cache.
        """
        CONFIGURATION['allow_cidr'] = [
            '192.30.252.0/22',
            '108.171.174.178/32',
            ]
        expand_allowed_ips()

        self.assertFalse('192.30.251.255' in CONFIGURATION['_allowed_ips'])
        self.assertTrue('192.30.252.0' in CONFIGURATION['_allowed_ips'])
        self.assertTrue('192.30.254.34' in CONFIGURATION['_allowed_ips'])
        self.assertTrue('192.30.255.255' in CONFIGURATION['_allowed_ips'])
        self.assertFalse('192.30.256.0' in CONFIGURATION['_allowed_ips'])

        self.assertFalse('108.171.174.177' in CONFIGURATION['_allowed_ips'])
        self.assertTrue('108.171.174.178' in CONFIGURATION['_allowed_ips'])
        self.assertFalse('108.171.174.179' in CONFIGURATION['_allowed_ips'])

    def test_hook_bad_ip(self):
        """
        An error is raised when request is comming from an untrusted IP.
        """
        request = DummyRequest([])
        request.client = IPv4Address('TCP', '111.0.0.0', 2345)
        result = hook(request, 'hook_name')

        self.assertEqual(403, request.code)
        self.assertTrue(result.startswith('Error:001: '))

    def test_hook_no_header(self):
        """
        An error is raised when X-Github-Event header is not present.
        """
        result = hook(self.request, 'hook_name')

        self.assertTrue(result.startswith('Error:004: '))

    def test_hook_bad_content_type(self):
        """
        An error is raised when the request has an unsuperted content type.
        """
        self.request.headers['x-github-event'] = 'push'
        self.request.headers['content-type'] = 'unknown'

        result = hook(self.request, 'hook_name')

        self.assertTrue(result.startswith('Error:002: '))

    def test_hook_bad_content(self):
        """
        An error is raised on parser failures.
        """
        self.request.headers['x-github-event'] = 'push'
        self.request.headers['content-type'] = (
            'application/x-www-form-urlencoded')
        self.request.content = StringIO('bad-json-formtat')

        result = hook(self.request, 'hook_name')

        self.assertTrue(result.startswith('Error:003: '))

    def test_hook_all_good_application_json(self):
        """
        Generic event having json content, are parsed.
        """
        self.called_event = None

        def callbacks(event):
            self.called_event = event

        CONFIGURATION['callback'] = callbacks
        self.request.headers['x-github-event'] = 'issue_comment'
        self.request.headers['content-type'] = 'application/json'
        self.request.content = StringIO('{"key": "value"}')

        result = hook(self.request, 'hook_name')

        self.assertIsNone(result)
        self.assertIsNotNone(self.called_event)
        self.assertEqual('hook_name', self.called_event.hook)
        self.assertEqual('issue_comment', self.called_event.name)
        self.assertEqual({'key': 'value'}, self.called_event.content)

    def test_hook_all_good_form_urlencoded(self):
        """
        x-www-form-urlencoded are parsed based on "payload" argument.
        """
        self.called_event = None

        def callbacks(event):
            self.called_event = event
        CONFIGURATION['callback'] = callbacks
        self.request.headers['x-github-event'] = 'push'
        self.request.headers['content-type'] = (
            'application/x-www-form-urlencoded')
        self.request.args = {'payload': ['{"key": "value"}']}

        result = hook(self.request, 'hook_name')

        self.assertIsNone(result)
        self.assertIsNotNone(self.called_event)
        self.assertEqual('hook_name', self.called_event.hook)
        self.assertEqual('push', self.called_event.name)
        self.assertEqual({'key': 'value'}, self.called_event.content)


txghserf/cidr.py
#!/usr/bin/python
# Copyright (c) 2007 Brandon Sterne
# Licensed under the MIT license.
# http://brandon.sternefamily.net/files/mit-license.txt
# CIDR Block Converter - 2007
import re


# convert an IP address from its dotted-quad format to its
# 32 binary digit representation
def ip2bin(ip):
    b = ""
    inQuads = ip.split(".")
    outQuads = 4
    for q in inQuads:
        if q != "":
            b += dec2bin(int(q), 8)
            outQuads -= 1
    while outQuads > 0:
        b += "00000000"
        outQuads -= 1
    return b


# convert a decimal number to binary representation
# if d is specified, left-pad the binary number with 0s to that length
def dec2bin(n, d=None):
    s = ""
    while n > 0:
        if n & 1:
            s = "1" + s
        else:
            s = "0" + s
        n >>= 1
    if d is not None:
        while len(s) < d:
            s = "0" + s
    if s == "":
        s = "0"
    return s


# convert a binary string into an IP address
def bin2ip(b):
    ip = ""
    for i in range(0, len(b), 8):
        ip += str(int(b[i:i + 8], 2)) + "."
    return ip[:-1]


# input validation routine for the CIDR block specified
def validateCIDRBlock(b):
    # appropriate format for CIDR block ($prefix/$subnet)
    p = re.compile("^([0-9]{1,3}\.){0,3}[0-9]{1,3}(/[0-9]{1,2}){1}$")
    if not p.match(b):
        print "Error: Invalid CIDR format!"
        return False
    # extract prefix and subnet size
    prefix, subnet = b.split("/")
    # each quad has an appropriate value (1-255)
    quads = prefix.split(".")
    for q in quads:
        if (int(q) < 0) or (int(q) > 255):
            print "Error: quad " + str(q) + " wrong size."
            return False
    # subnet is an appropriate value (1-32)
    if (int(subnet) < 1) or (int(subnet) > 32):
        print "Error: subnet " + str(subnet) + " wrong size."
        return False
    # passed all checks -> return True
    return True


def get_IP_list(block):
    """
    Return IP list address from block.
    """
    parts = block.split("/")
    base_ip = ip2bin(parts[0])
    subnet = int(parts[1])

    # If a subnet of 32 was specified simply return the single IP.
    if subnet == 32:
        return [bin2ip(base_ip)]

    result = []
    net = (32 - subnet)
    ip_prefix = base_ip[:-net]
    for i in range(2 ** net):
        ip = bin2ip(ip_prefix + dec2bin(i, net))
        result.append(ip)
    return result


def is_IP_in_block(ip, block):
    """
    Return True if IP belongs to CIDR block.
    """
    for member_ip in get_IP_list(block):
        if member_ip == ip:
            return True
    return False


txghserf/static/js/app.js
/*
Simple JS Application to manage GitHub Hooks.
*/

var app = angular.module('githubhooks', ['ui.bootstrap']).
  config(['$routeProvider', function($routeProvider) {
    $routeProvider.
      when('/login', {
        templateUrl: 'static/partials/login.html',
        controller: LoginCtrl
        }).
      when('/organizations', {
        templateUrl: 'static/partials/organizations.html',
        controller: OrganizationsCtrl
        }).
      when('/organization/:organization', {
        templateUrl: 'static/partials/organization.html',
        controller: OrganizationCtrl
        }).
      when('/organization/:organization/:repository', {
        templateUrl: 'static/partials/repository.html',
        controller: RepositoryCtrl
        }).
      otherwise({
        redirectTo: '/login'
    })
  }]).
  controller('rootCtrl', RootCtrl).
  factory('github', GitHubService)


function GitHubService($http, $location){
  var exports = {}

  exports.API_URL = 'https://api.github.com';

  var parseRelativeLinks = function(raw_links){
    var result= {}
    var members = raw_links.split(',')
    angular.forEach(
      members,
      function(value, key){
        var match = value.match('"*<(.*)>; rel="(\\w+)"')
        if (match) {
          result[match[2]] = match[1]
        }

      })
    return result
  }

  exports.GitHub = function(config){
      this.username = config.username
      this.password = config.password

      this.headers = {
        'Accept': 'application/vnd.github.raw',
        'Content-Type': 'application/json',
        'Authorization': 'Basic ' + Base64.encode(
          this.username + ':' + this.password)
      }

      this.get = function(url) {
        var self = this
        return $http({
          method: 'GET',
          url: url,
          cache: true,
          params: {per_page: 100},
          withCredentials: true,
          headers: self.headers
          })
      }

      this.post = function(url, data) {
        return $http({
          method: 'POST',
          data: data,
          url: url,
          withCredentials: true,
          headers: this.headers
          })
      }

      this.del = function(url) {
        return $http({
          method: 'DELETE',
          url: url,
          withCredentials: true,
          headers: this.headers
          })
      }

      this.patch = function(url, data) {
        return $http({
          method: 'PATCH',
          url: url,
          data: data,
          withCredentials: true,
          headers: this.headers
          })
      }

      return this
    }

  return exports
}

function RootCtrl($scope, $location) {

  $scope.github = ''
  $scope.account = ''
  $scope.alert = {}
  $scope.alert.info = ''
  $scope.alert.error = ''

  $scope.base_url = (
    $location.protocol() + '://' +
    $location.host() + ':' + $location.port() + '/' +
    'hook/'
    )

  $scope.getGitHub = function() {
    if (!$scope.github || !$scope.account) {
      $location.path('/')
      return
    }
    return $scope.github
  }

  $scope.goToLogin = function() {
    $location.path('/')
  }

  $scope.goToOrganizations = function() {
    $location.path('/organizations')
  }

  $scope.goToOrganization = function(organization) {
    $location.path('/organization/' + organization.login)
  }

  $scope.getOrganization = function(name) {
    var orgs = $scope.account.organizations
    var length = orgs.length
    for (var i = 0; i < length; i++) {
      if (orgs[i].login == name) {
        return orgs[i]
      }
    }
  }

  $scope.goToRepository = function(organization, repository) {
    $location.path(
      '/organization/' + organization.login + '/' + repository.name)
  }

  $scope.getRepository = function(organization, repository_name) {
    var repos = organization.repositories
    var length = repos.length
    for (var i = 0; i < length; i++) {
      if (repos[i].name == repository_name) {
        return repos[i]
      }
    }
  }

}

/*
Login handler.
*/
function LoginCtrl($scope, $rootScope, github) {

  $scope.user = {}
  $scope.user.name = ''
  $scope.user.password = ''

  $scope.doLogin = function() {
    $scope.$parent.alert.info = 'Authentication in progress...'


    hub = new github.GitHub({
      username: $scope.user.name,
      password: $scope.user.password
    })
    $scope.$parent.github = hub

    hub.get(github.API_URL + '/user').
      success(function(data, status) {
        $scope.$parent.alert.info = ''
        $scope.$parent.alert.error = ''
        $scope.$parent.account = data
        $scope.$parent.goToOrganizations()
      }).
      error(function(data, status) {
        $scope.$parent.alert.info = ''
        $scope.$parent.alert.error = data.message
        $scope.$parent.goToLogin()
      })
  }
}

/*
Organizations for an account.
*/
function OrganizationsCtrl($scope) {

  var hub = $scope.$parent.getGitHub()
  if (!hub) return

  hub.get($scope.$parent.account.organizations_url).
    success(function(data, status, headers){
      $scope.$parent.account.organizations = data
    }).
    error(function(data, status){
      $scope.$parent.error = data.message
      $scope.$parent.goToLogin()
    })

}


/*
Repos for an organization.
*/
function OrganizationCtrl($scope, $routeParams) {

  var hub = $scope.$parent.getGitHub()
  if (!hub) return

  $scope.name = $routeParams.organization
  $scope.organization = $scope.$parent.getOrganization($scope.name)

  hub.get($scope.organization.repos_url).
    success(function(data, status){
      $scope.organization.repositories = data
    }).
    error(function(data, status){
      $scope.$parent.error = data.message
      $scope.$parent.goToLogin()
    })
}


/*
Hooks for a repository.
*/
function RepositoryCtrl($scope, $routeParams, $dialog) {

  $scope.alerts = []

  $scope.addError = function(message) {
    $scope.alerts.push({type: 'error', msg: message})
  }

  $scope.addInfo = function(message) {
    $scope.alerts.push({type: 'info', msg: message})
  }

  $scope.addSuccess = function(message) {
    $scope.alerts.push({type: 'success', msg: message})
  }

  $scope.closeAlert = function(index) {
    $scope.alerts.splice(index, 1)
  }

  $scope.clearAlert = function(index) {
    $scope.alerts.length = 0
  }

  $scope.canEdit = function(hook) {
    if (hook.name === 'web') {
      return true
    } else {
      return false
    }
  }

  $scope.getStatus = function(hook) {
    var result
    if (hook.active) {
      result = 'active'
    } else {
      result = 'disabled'
    }
    return result
  }

  /*
  Return the name of the hook.
  */
  $scope.getName = function(hook) {
    if (hook.name === 'web') {
      return hook.config.url
    } else {
      return hook.name
    }
  }

  /*
  Simple view of hook configuration.
  */
  $scope.onView = function(hook){
    var buttons = [
      {result:'ok', label: 'OK', cssClass: 'btn-primary'}]

    $dialog.messageBox(
      'Details for ' + hook.name,
      hook.config,
      buttons
      ).open()
  }

  /*
  Action pefromed when add hook is requested from GUI.
  */
  $scope.onAdd = function(){

    var options = {
      backdrop: true,
      keyboard: true,
      backdropClick: false,
      resolve: {
        base_url: $scope.$parent.base_url,
        hook: undefined
      },
      templateUrl:  '/static/partials/hook.html',
      controller: 'HookCtrl'
    };

    $dialog.dialog(options).open().then(function(new_hook){
      if (new_hook) {
        $scope._addHook(new_hook)
      }
    })
  }

  $scope._addHook = function(new_hook){
    $scope.addInfo('Adding new hook...')
    hub.post($scope.repository.hooks_url, new_hook).
      success(function(data, status){
        $scope.clearAlert()
        $scope.addSuccess('New Hook added.')
        $scope.repository.hooks.push(data)
      }).
      error(function(data, status){
        $scope.clearAlert()
        $scope.addError(data)
      })
  }

  /*
  Action pefromed when edit hook is requested from GUI.
  */
  $scope.onEdit = function(hook){

    var options = {
      backdrop: true,
      keyboard: true,
      backdropClick: false,
      resolve: {
        base_url: $scope.$parent.base_url,
        hook: angular.copy(hook)
        },
      templateUrl:  '/static/partials/hook.html',
      controller: 'HookCtrl'
    };

    $dialog.dialog(options).open().then(function(result){
      if(result) {
        $scope._editHook(hook, result)
      }
    })
  }

  /*
  Perform hook editing.
  */
  $scope._editHook = function(hook, update) {
    $scope.addInfo('Updating hook ' + hook.id)
    hub.patch(hook.url, update).
      success(function(data, status){
        $scope.clearAlert()
        $scope.addSuccess('Hook updated')
        angular.extend(hook, data)
      }).
      error(function(data, status){
        $scope.clearAlert()
        $scope.addError(data)
      })
  }

  /*
  Action performed when delete button was pressed.
  */
  $scope.onDelete = function(hook){
    var title = 'Delete hook - ' + hook.name
    var msg = 'All hook inforamation will be lost.'
    var buttons = [
      {result:'cancel', label: 'Cancel'},
      {result:'delete', label: 'OK', cssClass: 'btn-danger'}]

    $dialog.messageBox(title, msg, buttons)
      .open()
      .then(function(result){
        if (result !== 'delete') {
          return
        }
        $scope._deleteHook(hook)
    })
  }

  /*
  Delete a hook.
  */
  $scope._deleteHook = function(hook){
    $scope.addInfo('Removing hook ' + hook.id)
    hub.del(hook.url).
      success(function(data, status){
        angular.forEach(
          $scope.repository.hooks,
          function(value, key){
            if (hook.id === value.id) {
              $scope.repository.hooks.splice(key, 1)
              return
            }
          })
        $scope.clearAlert()
        $scope.addSuccess('Hook removed.')
      }).
      error(function(data, status){
        $scope.clearAlert()
        $scope.addError(data)
      })
  }

  $scope.onTest = function(hook){
    var title = 'Test for ' + hook.name
    var buttons = [
      {result:'ok', label: 'OK', cssClass: 'btn-primary'}]

    hub.post(hook.test_url).
      success(function(data, status){
        $dialog.messageBox(
          title,
          'Test sent',
          buttons
          ).open()
      }).
      error(function(data, status){
        $dialog.messageBox(
          title,
          'Failed: ' + status + '-' + data,
          buttons
          ).open()
      })
  }

  var hub = $scope.$parent.getGitHub()
  if (!hub) return

  $scope.organization = $scope.$parent.getOrganization(
    $routeParams.organization)
  $scope.repository = $scope.$parent.getRepository(
    $scope.organization, $routeParams.repository)

  hub.get($scope.repository.hooks_url).
    success(function(data, status){
      $scope.repository.hooks = data
    }).
    error(function(data, status){
      $scope.$parent.error = data.message
      $scope.$parent.goToLogin()
    })

}


/*
Controller required for Hook add dialog.
*/
function HookCtrl($scope, dialog, base_url, hook){

  $scope.hook = {}
  $scope.hook.config = {}
  $scope.hook.config.content_type = 'form'

  $scope.base_url = base_url
  $insecure = true

  if (hook) {
    angular.extend($scope.hook, hook)
    $scope.title = "Edit hook " + hook.id
    $scope.action_name = "Save"
    if ($scope.hook.config.insecure_ssl == '1') {
      $scope.hook.config.insecure_ssl = true
    } else {
      $scope.hook.config.insecure_ssl = false
    }
  } else {
    $scope.title = "Add new hook"
    $scope.hook.name = 'web'
    $scope.hook.active = true
    $scope.hook.config.url = base_url
    $scope.action_name = "Add new hook"
  }

  $scope.close = function(result){
    dialog.close(result);
  }

  $scope.doSubmit = function() {
    if (!angular.isArray($scope.hook.events)){
      $scope.hook.events = $scope.hook.events.split(',')
      angular.forEach(
        $scope.hook.events,
        function(value, key) {
          $scope.hook.events[key] = $scope.hook.events[key].replace(/^\s+|\s+$/g, '')
        })
    }

    if ($scope.hook.config.insecure_ssl) {
      $scope.hook.config.insecure_ssl = '1'
    } else {
      $scope.hook.config.insecure_ssl = '0'
    }

    dialog.close($scope.hook);
  }
}


txghserf/server.py
"""
This is the part where requests are dispached.
"""
import os

try:
    import json
    # Shut up the linter.
    json
except ImportError:
    import simplejson as json

from klein import resource, route
from twisted.python import log
from twisted.web.static import File

from cidr import get_IP_list

# Shut up the linter.
resource

CONFIGURATION = {
    # Update it from https://api.github.com/meta
    'allow_cidr': [
        '127.0.0.1/24',
        "207.97.227.253/32",
        "50.57.128.197/32",
        "108.171.174.178/32",
        "50.57.231.61/32",
        "204.232.175.64/27",
        "192.30.252.0/22",
        ],
    'callback': None,

    # Cached list of allowed IP address.
    # Expanded at startup based on CIDR.
    # call `expand_allowed_ips()` if `allow_cidr` is changed at runtime.
    '_allowed_ips': {},
}

STATIC_PATH = os.path.join(os.path.dirname(__file__), 'static')


def expand_allowed_ips():
    """
    Expand the cached list of allowed ips.
    """
    CONFIGURATION['_allowed_ips'] = {}

    for block in CONFIGURATION['allow_cidr']:
        for ip in get_IP_list(block):
            CONFIGURATION['_allowed_ips'][ip] = True

# Do initial expansion.
expand_allowed_ips()


class Event(object):
    """
    Simple container for GitHub Event.
    """
    def __init__(self, hook, name, content):
        self.hook = hook
        self.name = name
        self.content = content

    def __str__(self):
        return """
        hook: %(hook)s
        event: %(event)s
        content:\n%(content)s
            """ % {
        'hook': self.hook,
        'event': self.name,
        'content': self.content,
        }


class ServerException(Exception):
    """
    Generic server exception.
    """
    def __init__(self, message):
        self.message = message


@route('/')
def root(request):
    """
    Entry point.
    """
    return File(STATIC_PATH)


@route('/static', branch=True)
def static(request):
    """
    All files from static.
    """
    return File(STATIC_PATH)


@route('/ping',  methods=['GET'])
def ping(request):
    """
    Simple resource to check that server is up.
    """
    return 'pong'


@route('/hook/<string:hook_name>',  methods=['POST'])
def hook(request, hook_name):
    """
    Main hook entry point.

    Check that request is valid, parse the content and then pass
    the object for further processing.
    """
    if request.getClientIP() not in CONFIGURATION['_allowed_ips']:
        request.code = 403
        log.msg(
            'Receive request for hook "%(name)s" from unauthorized'
            ' "%(ip)s".' % {'name': hook_name, 'ip': request.getClientIP()})
        return "Error:001: Where are you comming from?"

    event_name = request.getHeader('X-Github-Event')
    if not event_name:
        log.msg('No event name for "%(name)s". %(details)s' % {
                'name': hook_name, 'details': request.headers.items()})
        return "Error:004: What event is this?"

    content = None
    try:
        content = parse_request(request, event_name)
    except ServerException, error:
        log.msg('Failed to get json for hook "%(name)s". %(details)s' % {
                'name': hook_name, 'details': error.message})
        return "Error:002: Failed to get hook content."
    except:
        import traceback
        log.msg(
            'Failed to process "%(hook_name)s" "%(event_name)s":\n'
            '%(content)s\n'
            '%(details)s' % {
                'hook_name': hook_name,
                'event_name': event_name,
                'content': content,
                'details': traceback.format_exc(),
                })
        return "Error:003: Internal error"

    event = Event(
        hook=hook_name,
        name=event_name,
        content=content,
        )

    callback = CONFIGURATION['callback']
    if callback:
        callback(event)
    else:
        log.msg("Received new event for %s" % event)


def parse_request(request, event_name):
    """
    Return the event name nad JSON from request.
    """

    SUPPORTED_CONTENT_TYPES = [
        'application/x-www-form-urlencoded',
        'application/json',
        ]

    content_type = request.getHeader('Content-Type')
    if not content_type or content_type not in SUPPORTED_CONTENT_TYPES:
        raise ServerException('Unsuported content type.')

    if content_type == 'application/json':
        json_serialization = request.content.getvalue()
    elif content_type == 'application/x-www-form-urlencoded':
        json_serialization = request.args['payload'][0]
    else:
        raise AssertionError('How did we get here?')

    json_dict = json.loads(json_serialization)
    return json_dict
